### PR TITLE
Mjuric/lancaster stemmer rules file

### PR DESCRIPTION
**Pull request by mjunsilo**
#1001 - Lancaster Stemmer custom rules configuration

The latest SMILE library 1.22 is available on Maven Central now, which includes a custom rules feature for the Lancaster stemmer. I integrated a rudimentary rules loader from a file or resource, but I am not sure its the proper way to do it, since I am not familiar with the standard way to do this in DKPro. I'll change the code if someone could point me in the right direction.

### CODE REVIEW

**Comment by reckart** [comment]
We usually use `ResourceUtils.resolveLocation` for such cases. Cf. the `PdfReader` class for loading the *substitution table* or uses in `ResourceObjectProviderBase`.

```
                URL url = ResourceUtils.resolveLocation(substitutionTableLocation, this, aContext);
```

**Comment by mjunsilo** [comment]
Thanks. I have now changed it to use ResourceUtils during initialization. Let me know if there are other changes required.

File: dkpro-core-lancaster-asl/src/main/java/org/dkpro/core/lancaster/LancasterStemmer.java [hunk 1]
```diff
@@ -55,6 +59,15 @@
     @ConfigurationParameter(name = PARAM_STRIP_PREFIXES, mandatory = false, defaultValue = "false")
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/org/dkpro/core/lancaster/LancasterStemmer.java]: A primitive boolean parameter must have set `mandatory` to `true` as `null` is not a valid value.

File: dkpro-core-lancaster-asl/src/main/java/org/dkpro/core/lancaster/LancasterStemmer.java [hunk 2]
```diff
@@ -55,12 +62,51 @@
...
     private boolean stripPrefix;
 
+    /**
+     * Specifies an URL that should resolve to a location from where to load custom rules. If the location starts with
+     * {@code classpath:} the location is interpreted as a classpath location, e.g. "classpath:my/path/to/the/rules".
+     * Otherwise it is tried as an URL, file and at last UIMA resource.
+     * @see ResourceUtils
+     */
+    public static final String PARAM_CUSTOM_RULES = "customRules";
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/org/dkpro/core/lancaster/LancasterStemmer.java]: For this component, the rules serve as a model. Hence, we'd use PARAM_MODEL_LOCATION here:

```
	public static final String PARAM_MODEL_LOCATION = ComponentParameters.PARAM_MODEL_LOCATION;
	@ConfigurationParameter(name = PARAM_MODEL_LOCATION, mandatory = false)
	protected String modelLocation;
```

**Comment by mjunsilo** [comment]
I could remove the explicit english language requirements when a custom model is defined, but I think it would then also make sense to tie the model to a language. What are your thoughts on this, and how do you normally bundled them in a configuration?

**Comment by reckart** [comment]
If a component supports models, we have two options:

* **via language and variant** - the developer packages up models as Maven JARs and adds them to the `managedDependencies`section of the component's POM. The users then uses `PARAM_LANGUAGE` (or the language from the CAS) and `PARAM_VARIANT` to control which model they want to use. The developer may define a all-language default variant set up a *default variants* file that determines the default variant per-language, such that the user doesn't have to specify a variant. The packaging of such models happens usually through a `build.xml` file (cf. `src/scripts` folder in the OpenNLP, CoreNLP, or UDPipe modules).

* **via `PARAM_MODEL_LOCATION` - in this case, the `PARAM_LANGUAGE` and `PARAM_VARIANT` are (usually) ignored. The model is then loaded from a file system or classpath (i.e. JAR) location. This is usually only needed if a user has a self-trained model or custom rules file. 

I believe most users use the models that are registered with DKPro Core, so they use the first of the described approaches. Model files declare the language they support in their metadata and the DKPro Core documentation generator picks it up from there. Components making use of such models do not declare the language capability.

**Comment by mjunsilo** [comment]
Thanks. The custom model feature is mostly intended for those that want to tweak the default rules, and given that only english is supported by default, then option one is pretty much out of the question. I also don't know if anyone has done Lancaster stemming rules for other languages. However, I can easily support the PARAM_LANGUAGE parameter for those very rare cases, just for the sake of completeness. They would still have to implement their own flow control to switch model dynamically based on document language.

**Comment by reckart** [comment]
@mjunsilo btw, I don't know if you noticed - you have commit permissions to the repo with the ICLA being on file and since you've already contributed a few things. I would suggest you just commit and we review post-commit instead of via PRs. What do you think?

**Comment by mjunsilo** [comment]
Sure, I'll try that the next time.

BTW Sorry, I forgot to add the capability annotations, but thanks for doing it for me.

**Pull request merged**